### ISSUE TITLE
Proposal: emacs-like key-bindings and key sequence support

### ISSUE THREAD

**Issue by ChaosEternal** [amirouche/zk#4]
I am implementing an emacs-like key-binding and key sequence support.
I will add an proc:

 `%bind-key`

for example 
```scheme
(%bind-key `(,TB-KEY-CTRL-X ,TB-KEY-CTRL-C) quit)
```
 will bind  key sequence ctrl-x ctrl-c to procedure quit, that mean if a user press ctrl-x and ctrl-c  in sequence will make zk to quit.

`bind-key`
bind key-sequence in names to proc:
`(bind-key "c-x c-c" quit) `
which has the same effect as last example, but all key names are in emacs style.

**Comment by amirouche** [comment]
You are right using emacs-like key-bindings is a good choice.

The current implementation relies on strings to declare bindings, I don't like it because it requires to validate the string somehow to allow good enough user experience.

My "counter" proposal would be to use something like parser combinators to describe key bindings.

**Comment by amirouche** [comment]
For instance, I am thinking about the following API to bind `quit`:

```scheme
(key-bind quit (ctrl (key #\x)) (ctrl (key #\c)))
```

Then we can imagine other bindings like:

```scheme
(key-bind minibuffer-input (alt (key #\x)))
(key-bind undo (ctrl (shift (key #\z)))
``` 

`ctrl` `shift` `alt` and `key` would be procedure that validate their input and return procedure that might look like the following:

```scheme
(define proc
  ((case-lambda
    ;; return cons representation of the key binding for introspection need (help)
    (() '(ctrl (key z))
    ;; where event is one of `event-key` or `event-char` so we might support 
    ;; an :robot: mode :100: 
    ((event) (%dispatch event)))) 
```

WDYT?

**Comment by amirouche** [comment]
:robot: mode is my proposal for the :smiling_imp: evil mode of emacs

:wink:

**Comment by ChaosEternal** [comment]
No, M-x is actually working in my side, what's your version of Termbox?

in termbox, M-x is generating two events in sequential, the first event is
ESC, just like the ESC key is pressed, the second is #\x itself.

So in my implementation, this key bind is really supported, it's supported
the same way as other key sequences.

A quick example:

(bind-key "M-x" quit)
and
press M-x will call quit.

**Comment by amirouche** [comment]
Still `M-x` doesn't work on my side, I need to do something special to have `ESC` translate to `M-x` apparently but still, I am not sure termbox will support it.
